pytorch_lightning/trainer/data_loading.py

```diff
@@ -133,19 +138,21 @@ def auto_add_sampler(self, dataloader: DataLoader, shuffle: bool) -> DataLoader:
                 )
 
             # replace with distributed sampler
-            sampler = self._get_distributed_sampler(dataloader, shuffle)
-            dataloader = self.replace_sampler(dataloader, sampler)
+            sampler = self._get_distributed_sampler(dataloader, shuffle, mode=mode)
+            dataloader = self.replace_sampler(dataloader, sampler, mode=mode)
 
         return dataloader
 
     @staticmethod
-    def _resolve_batch_sampler(dl_args, dataloader, sampler):
+    def _resolve_batch_sampler(dl_args, dataloader, sampler, mode: Optional[RunningStage] = None) -> Dict[str, Any]:
         batch_sampler = getattr(dataloader, "batch_sampler")
-        if batch_sampler is not None and type(batch_sampler) is not BatchSampler:
-            batch_sampler = type(batch_sampler)(
+        is_predicting = mode == RunningStage.PREDICTING
+        if (batch_sampler is not None and type(batch_sampler) is not BatchSampler) or mode == RunningStage.PREDICTING:
+            batch_sampler_type = IndexBatchSampler if is_predicting else type(batch_sampler)
```

Comment:
Silently replacing is dangerous. I see two options

- Print a warning if the provided batch_sampler is not the default
- Compose the samplers?

Reply:
Done !